cancer_ai/validator/rewarder.py

```diff
@@ -1,159 +1,193 @@
-                hotkey=hotkey,
-                leader_since=datetime.now(timezone.utc),
-                model_result=winner_model_result,
-            )
-            return
 
-        # Score degradation starts on 3rd week of leadership
-        base_share = 1 / len(self.competition_leader_mapping)
-        if days_as_leader > REWARD_REDUCTION_START_DAY:
-            periods = (
-                days_as_leader - REWARD_REDUCTION_START_DAY
-            ) // REWARD_REDUCTION_STEP_DAYS
-            reduction_factor = max(
-                REWARD_REDUCTION_STEP, 1 - REWARD_REDUCTION_STEP * periods
-            )
-            final_share = base_share * reduction_factor
-            reduced_share = base_share - final_share
-            return final_share, reduced_share
-        return base_share, 0
-
-    async def update_scores(
-        self,
-        winner_hotkey: str,
-        competition_id: str,
-        winner_model_result: ModelEvaluationResult,
-    ):
-        """
-        Update the scores of the competitors based on the winner of the competition.
-
-        Args:
-            winner_hotkey: Competition winner's hotkey.
-            competition_id: Competition ID.
-            winner_model_result: Information about the winner's model.
-
-        """
-        result_improved = False
-        # Logic to check if new winner's model made any improvement. If not, keep current winner
-        if (
-            len(self.competition_leader_mapping) > 0
-            and competition_id in self.competition_leader_mapping
-        ):
-            current_leader_model_result = self.competition_leader_mapping[
-                competition_id
-            ].model_result
+class ModelScore(BaseModel):
+    date: datetime
+    score: float
 
-            result_improved = (
-                winner_model_result.score - current_leader_model_result.score > 0.001
-            )
 
-            if not result_improved:
-                winner_hotkey = self.competition_leader_mapping[competition_id].hotkey
+class CompetitionResultsStore(BaseModel):
+    # Structure: {competition_id: {hotkey: [ModelScore, ...]}}
+    score_map: dict[str, dict[Hotkey, list[ModelScore]]] = {}
+    # Structure: {competition_id: {hotkey: average_score}}
+    average_scores: dict[str, dict[Hotkey, float]] = {}
+    # Structure: {competition_id: (hotkey, score)}
+    current_top_hotkeys: dict[str, tuple[Hotkey, float]] = {}
+
+    def add_score(self, competition_id: str, hotkey: Hotkey, score: float, date: datetime = None):
+        """Add a score for a specific hotkey in a specific competition."""
+
+        # Initialize competition dictionaries if they don't exist
+        if competition_id not in self.score_map:
+            self.score_map[competition_id] = {}
+        if competition_id not in self.average_scores:
+            self.average_scores[competition_id] = {}
+
+        # Initialize hotkey list if it doesn't exist
+        if hotkey not in self.score_map[competition_id]:
```

Comment:
just an code aesthetics idea - maybe we can enclose all this logic related to initialization of states into some separate function and name it nicely to make it more clean here?